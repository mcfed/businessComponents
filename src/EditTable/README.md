# EditTable 组件

> 可编辑表格 API 基于 antd 组件中的 Table 组件，拓展了两个属性 editComponent，editConfig

## Usage

```
editComponent: ()=> <Input placeholder='请输入名字'/>
editConfig:{
        rules: [{required:true, message: '请输入名字'},{validator: 自定义规则}]
  }
```

## Option

| 参数          | 说明                                                                                                                                                 | 类型      |
| ------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------- | --------- |
| editComponent | 用于传递渲染的编辑组件,传递组件为 antd 基础  表单组件，也可自定义，需遵循 antd 组件的包装                                                            |  funciton |
| editConfig    | 用于传递初始值，验证规则等配置,支持 antd 表单基本校验方法， 可[查看文档](https://ant.design/components/form-cn/),支持按照 antd from 书写的自定义校验 | object    |
