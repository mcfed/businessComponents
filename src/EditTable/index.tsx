import * as React from 'react';
import {Table, Button, message, Form} from 'antd';
import {FormComponentProps} from 'antd/lib/form';
import {ColumnProps} from 'antd/lib/table/interface';
import {WrappedFormUtils} from 'antd/lib/form/Form';
import {HTMLAttributes} from 'react';

const FormItem = Form.Item;
const EditableContext = React.createContext({} as WrappedFormUtils);

const EditableRow: React.FunctionComponent<EditableRowProps> = ({
  form,
  index,
  ...props
}) => (
  <EditableContext.Provider value={form}>
    <tr {...props} />
  </EditableContext.Provider>
);

interface EditableRowProps
  extends HTMLAttributes<HTMLTableRowElement>,
    FormComponentProps {
  index: number;
}

const EditableFormRow = Form.create()(EditableRow);

declare type EditTableMode = 'full' | 'row'; // 全表格编辑 | 单行编辑

interface Item {
  key: string;
  [propName: string]: any;
}

interface ColumnsItem<T> extends ColumnProps<T> {
  dataIndex: string;
  editingStatus: boolean;
  renderCol: (text: string, row: T, instance: Object) => React.ReactNode;
  [propName: string]: any;
}

// type ColumnsItemWithoutRender<T> = Omit<ColumnsItem<T>, 'render'>;

/**
 * columns：表格列配置，必须
 * data: 数据数组，必须
 * mode: 编辑模式，非必须，默认为单行编辑
 */
interface EditTableProps<T> {
  /**
   * columns：表格列配置
   */
  columns: ColumnsItem<T>[];
  /**
   * 数据数组
   */
  data: T[];
  /**
   * 编辑模式
   */
  mode: EditTableMode;
  onChange(data: T[]): void;
}

interface State<T> {
  data: T[];
  editingKey: string;
  keyList: string[];
  columns: ColumnsItem<T>[];
}

type HTMLElementEvent<T extends HTMLElement> = Event & {
  target: T;
  currentTarget: T;
  // probably you might want to add the currentTarget as well
  // currentTarget: T;
};

export default class EditTable<T extends Item> extends React.Component<
  EditTableProps<T>,
  State<T>
> {
  static defaultProps = {
    mode: 'row'
  };
  constructor(props: Readonly<EditTableProps<T>>) {
    super(props);
    this.state = {
      data: [],
      editingKey: '',
      keyList: [],
      columns: []
    };
  }
  componentDidMount() {
    /* istanbul ignore else */
    if (this.props.columns && this.props.columns.length > 0) {
      this.setState(
        {
          data: this.props.data === undefined ? [] : this.props.data,
          columns: [...this.props.columns, ...this.state.columns]
        },
        () => {
          let keyList = this.state.columns.map(c => c.dataIndex);
          this.setState({
            keyList
          });
        }
      );
    }
  }
  UNSAFE_componentWillReceiveProps(nextprops: Readonly<EditTableProps<T>>) {
    /* istanbul ignore else */
    if (this.props.data !== nextprops.data) {
      this.setState({
        data: nextprops.data
      });
    }
  }

  isEditing = (record: Item) => {
    // isEditing = <T extends unknown>(record: {key:T}) => {
    return record.key === this.state.editingKey;
  };

  edit = (key: string) => {
    if (this.state.editingKey !== '') {
      message.error('请先保存编辑项再进行其他编辑操作！');
      return false;
    }
    this.setState({editingKey: key});
    this.activeStatus();
  };

  // 双击td事件
  editColumn = (key: string) => {
    if (this.state.editingKey !== '') {
      message.error('请先保存编辑项再进行其他编辑操作！');
      return false;
    }
    this.setState({editingKey: key});
  };

  // changeColumnEditStatus = (record: Item, tdObject: any) => {
  //   this.editColumn(record.key);
  //   this.state.columns.forEach((item: any) => {
  //     /* istanbul ignore else */
  //     if (item.dataIndex === tdObject.dataIndex) {
  //       item.editingStatus = true;
  //     }
  //   });
  // };

  revertStatus() {
    // 恢复每一列的编辑状态，去除所有editingStatus
    this.state.columns.map(
      (item: ColumnsItem<T>) => (item.editingStatus = false)
    );
  }

  activeStatus() {
    // 激活每一列的编辑状态，所有列editingStatus设为true
    this.state.columns.map(
      (item: ColumnsItem<T>) => (item.editingStatus = true)
    );
  }

  delete(key: string) {
    let newData = [...this.state.data];
    this.setState(
      {
        data: newData.filter(c => c.key !== key),
        editingKey: ''
      },
      () => {
        this.props.onChange(this.state.data);
      }
    );
  }
  save(form: WrappedFormUtils, key: string) {
    form.validateFields((error: any, row: T) => {
      if (error) {
        return;
      }
      const newData = [...this.state.data];
      const index = newData.findIndex((item: Item) => key === item.key);
      if (index > -1) {
        const item = newData[index];
        newData.splice(index, 1, {
          ...item,
          ...row,
          key: item.key
        });
      } else {
        newData.push(row);
      }
      this.setState({data: newData, editingKey: ''}, () => {
        this.props.onChange(newData);
      });
    });
  }

  cancel = (form: WrappedFormUtils, key: string) => {
    let obj = this.state.data.filter(d => d.key === key)[0];
    let Bdelete = false;
    for (let b in obj) {
      if (obj[b] === '') {
        Bdelete = true;
        break;
      }
    }
    if (Bdelete) {
      this.delete(key);
    }
    this.setState({editingKey: ''});

    this.revertStatus();
  };

  addNew = () => {
    if (this.state.editingKey !== '') {
      message.error('请先保存编辑项再进行添加操作！');
      return false;
    }
    let key = new Date().valueOf() + '' + Math.floor(Math.random() * 10 + 1);
    let obj: any = {
      key: key
    };
    let keyList = [...this.state.keyList];
    /* istanbul ignore else */
    if (keyList.length > 1) {
      keyList.length = keyList.length - 1;
    }
    keyList.forEach(d => {
      obj[d] = '';
    });
    let data = [...this.state.data];
    data.push(obj);
    this.setState({
      data,
      editingKey: key
    });

    this.activeStatus();
  };

  renderCell(text: string, record: Item, cellConfig: ColumnsItem<T>) {
    const {dataIndex, editComponent, editConfig} = cellConfig;
    const instance = this;
    const {mode} = this.props;
    return (
      <EditableContext.Consumer>
        {(form: WrappedFormUtils) => {
          const {getFieldDecorator, setFieldsValue} = form;
          const component = editComponent(text, record, instance, form);
          return (
            <FormItem style={{margin: 0}}>
              {getFieldDecorator(dataIndex, {
                ...editConfig,
                initialValue:
                  record[dataIndex] === ''
                    ? editConfig && editConfig.initialValue
                    : record[dataIndex]
              })(
                React.createElement(component.type, {
                  ...component.props,
                  ...(mode !== 'row'
                    ? {
                        onChange: function(
                          e: HTMLElementEvent<HTMLInputElement>
                        ) {
                          // e is event
                          if (e.target) {
                            setFieldsValue({[dataIndex]: e.target.value});
                          } else {
                            setFieldsValue({[dataIndex]: e});
                          }
                          instance.save(form, record.key);
                        }
                      }
                    : {})
                })
              )}
            </FormItem>
          );
        }}
      </EditableContext.Consumer>
    );
  }

  render() {
    const {data, columns, mode, onChange, ...otherProps} = this.props;
    const components = {
      body: {
        row: EditableFormRow
        // cell: EditableCell
      }
    };
    const instance = this;

    const columnsFinal = this.state.columns.map((col: ColumnsItem<T>) => {
      if (col.editComponent === undefined) {
        return col;
      }
      return {
        ...col,
        // onCellClick: (record,index)=>{
        //   console.log(record,index)
        // },
        render: (text: string, row: T) => {
          return mode === 'full' || this.isEditing(row)
            ? this.renderCell(text, row, col)
            : // ts 中 render 为table clumns 关键字，用renderCol 替换
            // : col.render
            // ? col.render(text, row, instance)
            col.renderCol
            ? col.renderCol(text, row, instance)
            : text;
        }
      };
    });

    return (
      <Table
        components={components}
        bordered
        dataSource={this.state.data}
        columns={columnsFinal}
        rowClassName={(record: object, index: number) => 'editable-row'}
        {...otherProps}
        footer={() => (
          <Button icon='plus' onClick={this.addNew} style={{width: '100%'}}>
            新增
          </Button>
        )}
      />
    );
  }
}
