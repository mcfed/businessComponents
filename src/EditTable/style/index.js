import 'antd/es/table/style';
import 'antd/es/button/style';
import 'antd/es/message/style';
import 'antd/es/form/style';
