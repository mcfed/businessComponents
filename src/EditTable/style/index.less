.ant-table-tbody > .editable-row > td {
  padding: 20px 20px;
}
.editable-row {
  .ant-form-explain {
    position: absolute;
  }
  .ExtraColumn {
    .ant-form-explain {
      position: inherit;
    }
  }
}
