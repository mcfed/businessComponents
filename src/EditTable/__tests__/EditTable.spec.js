import {shallow} from 'enzyme';
import React from 'react';
import {Button} from 'antd';

import EditTable from '../index.tsx';

const setup = props => {
  const columns = [
    {
      title: '姓名',
      dataIndex: 'name',
      key: 'name'
    },
    {
      title: '年龄',
      dataIndex: 'age',
      key: 'age'
    },
    {
      title: '住址',
      dataIndex: 'address',
      key: 'address'
    }
  ];
  const dataSource = [
    {
      key: '1',
      name: '胡彦斌',
      age: 32,
      address: '西湖区湖底公园1号'
    },
    {
      key: '2',
      name: '胡彦祖',
      age: 42,
      address: '西湖区湖底公园1号'
    }
  ];

  const defaultProps = {
    onChange: jest.fn()
  };

  const wrapper = shallow(
    <EditTable
      data={dataSource}
      columns={columns}
      {...Object.assign({}, defaultProps, props)}
    />
  );

  return {
    wrapper,
    props,
    dataSource,
    columns
  };
};

describe.skip('edittable base test', () => {
  it('render ', () => {
    const {wrapper, props} = setup();
    // console.log(wrapper)
    expect(wrapper.find('Table').exists()).toBe(true);
  });

  it('默认state', () => {
    const {wrapper, props} = setup({
      data: undefined
    });

    expect(wrapper.state('data')).toEqual([]);
    expect(wrapper.state('editingKey')).toBe('');
  });

  it('传值后的state', () => {
    const {wrapper, props, dataSource} = setup();

    expect(wrapper.state('data')).toEqual(dataSource);
  });

  it('render 的columns 是重新加工过的', () => {
    const {wrapper, props, dataSource, columns} = setup();

    expect(wrapper.find('Table').prop('columns')).toEqual(
      wrapper.state('columns').map(col => {
        if (!col.editComponent) {
          return col;
        }
        return {
          ...col,
          onCell: record => ({
            record,
            editConfig: col.editConfig,
            editDom: col.editComponent,
            dataIndex: col.dataIndex,
            title: col.title,
            editing: this.isEditing(record)
          })
        };
      })
    );
  });

  it.skip('render footer', () => {
    const {wrapper, props} = setup();
    expect(wrapper.find('Table').prop('footer')()).toEqual(
      <Button
        icon='plus'
        onClick={wrapper.instance().addNew}
        style={{width: '100%'}}>
        新增
      </Button>
    );
  });
});

describe.skip('editable method ', () => {
  it('isEditing 默认为空，传入key为123 返回false', () => {
    const {wrapper, props} = setup();
    const record = {
      key: 123
    };
    expect(wrapper.instance().isEditing(record)).toBe(false);
  });

  it('isEditing 默认为空，传入key为空 返回true', () => {
    const {wrapper, props} = setup();
    const record = {
      key: ''
    };
    expect(wrapper.instance().isEditing(record)).toBe(true);
  });

  it('edit 方法调用,若key 为空，则返回false', () => {
    const {wrapper, props} = setup();
    const key = '';

    expect(wrapper.instance().edit(key)).toBe(undefined);
  });

  it('edit 方法调用，若key不为空，则 state 改变', () => {
    const {wrapper, props} = setup();

    const key = '123';
    wrapper.instance().edit(key);
    expect(wrapper.state('editingKey')).toBe('123');
  });

  it('delete 方法调用', () => {
    const onchangeFn = jest.fn();
    const {wrapper, props, dataSource} = setup({
      onChange: onchangeFn
    });
    const key = '1';
    let oldData = wrapper.state('data');
    wrapper.instance().delete(key);
    expect(wrapper.state('data')).toEqual(oldData.filter(c => c.key !== key));
    expect(onchangeFn.mock.calls.length).toBe(1);
  });

  it('addNew 方法调用,若当前editingkey不为空则返回false', () => {
    const {wrapper, props} = setup();
    wrapper.instance().edit('1');
    expect(wrapper.instance().addNew()).toBe(false);
  });

  it('addNew 方法调用', () => {
    const {wrapper, props, dataSource} = setup();
    wrapper.instance().addNew();
    expect(wrapper.state('data').length).toBe(dataSource.length + 1);
  });

  it.skip('cancel 方法调用', () => {
    const {wrapper, props} = setup();
    const params = {
      form: undefined,
      key: '1'
    };
    wrapper.instance().cancel(params);
    //判断 是否删除 逻辑？
  });

  it.skip('editable true branch 测试', () => {
    const {wrapper} = setup();
    const instance = wrapper.instance();
    const render = instance.state.columns.filter(
      item => item.dataIndex === '操作'
    )[0].render;
    const record = {
      key: '1'
    };
    instance.save = jest.fn();
    instance.isEditing = jest.fn(() => true);
    const renderWrapper = render('', record);
    // dom 点击事件覆盖
    renderWrapper.props.children.props.children[0].props
      .children()
      .props.onClick();
    renderWrapper.props.children.props.children[1].props
      .children()
      .props.onConfirm();
  });

  it.skip('editable false branch 测试', () => {
    const {wrapper} = setup();
    const instance = wrapper.instance();
    const render = instance.state.columns.filter(
      item => item.dataIndex === '操作'
    )[0].render;
    const record = {
      key: '1'
    };
    const renderWrapper = render('', record);
    // dom 点击事件覆盖
    renderWrapper.props.children.props.children[0].props.onClick();
    renderWrapper.props.children.props.children[1].props.onConfirm();
  });

  it('UNSAFE_componentWillReceiveProps 生命周期测试', () => {
    const {wrapper} = setup();
    const instance = wrapper.instance();
    const nextprops = {
      data: [
        {
          key: '3',
          name: '胡彦组',
          age: 32,
          address: '西湖区湖底公园1号'
        }
      ]
    };
    instance.UNSAFE_componentWillReceiveProps(nextprops);
  });

  it('edit 方法测试,editingKey已存在branch', () => {
    const {wrapper} = setup();
    const instance = wrapper.instance();
    const editingKey = '1';
    instance.state.editingKey = '5';
    const activeStatusSpy = jest.spyOn(instance, 'activeStatus');
    instance.edit(editingKey);
    expect(instance.state.editingKey).toBe('5');
    expect(activeStatusSpy).not.toHaveBeenCalledWith();
  });

  it('editColumn 方法测试', () => {
    const {wrapper} = setup();
    const instance = wrapper.instance();
    const editingKey = 1;
    instance.editColumn(editingKey);
    expect(instance.state.editingKey).toBe(editingKey);
  });

  it('editColumn 方法测试,editingKey已存在branch', () => {
    const {wrapper} = setup();
    const instance = wrapper.instance();
    const editingKey = '1';
    instance.state.editingKey = '5';
    instance.editColumn(editingKey);
    expect(instance.state.editingKey).not.toBe(editingKey);
  });

  // it('changeColumnEditStatus 方法测试', () => {
  //   const {wrapper} = setup();
  //   const instance = wrapper.instance();
  //   const record = {
  //     key: 1
  //   };
  //   const tdObject = {
  //     dataIndex: 'name'
  //   };
  //   const editColumnSpy = jest.spyOn(instance, 'editColumn');
  //   instance.changeColumnEditStatus(record, tdObject);
  //   expect(editColumnSpy).toHaveBeenCalledWith(record.key);
  //   expect(
  //     instance.state.columns.filter(
  //       item => item.dataIndex === tdObject.dataIndex
  //     )[0].editingStatus
  //   ).toBeTruthy();
  // });

  it('revertStatus 方法测试', () => {
    const {wrapper} = setup();
    const instance = wrapper.instance();
    instance.revertStatus();
    instance.state.columns.map(item => {
      expect(item.editingStatus).toBeFalsy();
    });
  });

  it('activeStatus 方法测试', () => {
    const {wrapper} = setup();
    const instance = wrapper.instance();
    instance.activeStatus();
    instance.state.columns.map(item => {
      expect(item.editingStatus).toBeTruthy();
    });
  });

  it('save 方法测试', () => {
    const {wrapper} = setup();
    const instance = wrapper.instance();
    const form = {
      validateFields: jest.fn(cb => {
        cb(false, null);
      }),
      getFieldInstance: jest.fn().mockReturnValue({
        validateFields: jest.fn(cb => {
          cb(false, null);
        })
      })
    };
    const key1 = '1';
    const key2 = '110';
    // 编辑保存
    instance.save(form, key1);
    // 保存新增项branch
    instance.save(form, key2);
  });

  it('save 方法测试,第一个验证方法未通过branch', () => {
    const {wrapper} = setup();
    const instance = wrapper.instance();
    const form = {
      validateFields: jest.fn(cb => {
        cb(true, null);
      })
    };
    const key = '1';
    instance.save(form, key);
  });

  it('save 方法测试,第二个验证方法未通过branch', () => {
    const {wrapper} = setup();
    const instance = wrapper.instance();
    const form = {
      validateFields: jest.fn(cb => {
        cb(false, null);
      }),
      getFieldInstance: jest.fn().mockReturnValue({
        validateFields: jest.fn(cb => {
          cb(true, null);
        })
      })
    };
    const key = '1';
    instance.save(form, key);
  });

  it('cancel 方法测试, 取消table的数据行branch', () => {
    const {wrapper} = setup();
    const instance = wrapper.instance();
    const deleteSpy = jest.spyOn(instance, 'delete');
    const revertStatusSpy = jest.spyOn(instance, 'revertStatus');
    instance.cancel({}, '1');
    expect(deleteSpy).not.toBeCalledWith('1');
    expect(instance.state.editingKey).toBe('');
    expect(revertStatusSpy).toBeCalled();
  });

  it('cancel 方法测试, 取消新增的数据行branch', () => {
    const {wrapper} = setup();
    const instance = wrapper.instance();
    const deleteSpy = jest.spyOn(instance, 'delete');
    const revertStatusSpy = jest.spyOn(instance, 'revertStatus');
    instance.state.data.push({
      key: '3',
      name: '',
      age: '',
      address: ''
    });
    instance.cancel({}, '3');
    expect(deleteSpy).toBeCalledWith('3');
    expect(instance.state.editingKey).toBe('');
    expect(revertStatusSpy).toBeCalled();
  });

  it('editComponent 存在情况测试', () => {
    const props = {
      columns: [
        {
          title: '姓名',
          dataIndex: 'name',
          key: 'name545555',
          editComponent: <div></div>
        },
        {
          title: '年龄',
          dataIndex: 'age',
          key: 'age'
        },
        {
          title: '住址',
          dataIndex: 'address',
          key: 'address'
        }
      ]
    };
    setup(props);
  });
});

describe('快照测试', () => {
  it('Editable全页快照', () => {
    const {wrapper} = setup();

    expect(wrapper).toMatchSnapshot();
  });
});
