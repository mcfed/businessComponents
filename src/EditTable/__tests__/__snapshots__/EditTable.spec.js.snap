// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`快照测试 Editable全页快照 1`] = `
<withStore(Table)
  bordered={true}
  columns={
    Array [
      Object {
        "dataIndex": "name",
        "key": "name",
        "title": "姓名",
      },
      Object {
        "dataIndex": "age",
        "key": "age",
        "title": "年龄",
      },
      Object {
        "dataIndex": "address",
        "key": "address",
        "title": "住址",
      },
    ]
  }
  components={
    Object {
      "body": Object {
        "row": [Function],
      },
    }
  }
  dataSource={
    Array [
      Object {
        "address": "西湖区湖底公园1号",
        "age": 32,
        "key": "1",
        "name": "胡彦斌",
      },
      Object {
        "address": "西湖区湖底公园1号",
        "age": 42,
        "key": "2",
        "name": "胡彦祖",
      },
    ]
  }
  footer={[Function]}
  rowClassName={[Function]}
/>
`;
